### PR TITLE
Parser inference also based on filename and version

### LINKED ISSUE #483: Couldn't resolve parser "json-stringify"
When trying to format json files I am getting this error. 

This happens in our monorepo when attempting to format any JSON file. Repo is here: https://github.com/forcedotcom/salesforcedx-vscode

Prettier VSCode Version 1.4.0
VS Code Version: 1.23.1
Node Version: v8.11.2
NPM Version: 6.0.1

### PR DESCRIPTION
**Pull request by CiGit**
Infer parser by taking into account version and filename.

Change tests, let prettier infer the parser based on filename.
Fix #483

### CODE REVIEW

File: src/PrettierEditProvider.ts [hunk 1]
```diff
@@ -35,21 +35,24 @@ async function hasPrettierConfig(filePath: string) {
     return config !== null;
 }
 
-type ResolveConfigResult = { config: PrettierConfig | null, error?: Error };
+type ResolveConfigResult = { config: PrettierConfig | null; error?: Error };
```
Comment (RobinMalfait) [inline src/PrettierEditProvider.ts]: Is this change correct? If so, than I should do some more research on typescript.

File: src/types.d.ts [hunk 2]
```diff
@@ -7,6 +7,7 @@ export type ParserOption =
     | 'scss'
     | 'typescript'
     | 'json'
+    | 'json-stringify'
```
Comment (RobinMalfait) [inline src/types.d.ts]: That makes sense 👍

File: src/PrettierEditProvider.ts (hunk 1 above)
Comment (CiGit) [reply src/PrettierEditProvider.ts]: Prettier did it.
interface or object type can separate their properties with `,` `;` or `\n`

Comment (RobinMalfait) [reply src/PrettierEditProvider.ts]: Interesting!

File: src/utils.ts [hunk 3]
```diff
@@ -12,10 +13,17 @@ export function getConfig(uri?: Uri): PrettierVSCodeConfig {
...
-    const language = getSupportLanguages().find(lang =>
-        lang.vscodeLanguageIds.includes(languageId)
+    const language = getSupportLanguages(version).find(
+        lang =>
+            lang.vscodeLanguageIds.includes(languageId) &&
+            // Only for some specific filenames
+            (lang.extensions.length > 0 ||
+                (path != null &&
+                    lang.filenames != null &&
+                    lang.filenames.includes(basename(path))))
```
Comment (CiGit) [inline src/utils.ts]: Note: this works because 'json-stringify' is before 'json' in the array.

Comment (CiGit) [reply src/utils.ts]: Should do better...

**Pull request merged**